From fb044c62b60fe5533d3157c6cbaeff4463bb3dcb Mon Sep 17 00:00:00 2001
From: "GONG, Ruiqi" <gongruiqi1@huawei.com>
Date: Tue, 11 Oct 2022 19:19:57 +0800
Subject: [PATCH 0593/1076] [Huawei] selinux: genfscon: add a config for this
 feature

Offering: HULK
hulk inclusion
category: feature
bugzilla: 186697

----------------------------------------

An option CONFIG_SECURITY_SELINUX_GENFSCON_NS is added to control the
enabling of the SELinux Genfscon Label Namespace feature in compilation.

Signed-off-by: GONG, Ruiqi <gongruiqi1@huawei.com>
---
 arch/arm64/configs/openeuler_defconfig |   1 +
 arch/x86/configs/openeuler_defconfig   |   1 +
 security/security.c                    |   4 +
 security/selinux/Kconfig               |  12 ++
 security/selinux/hooks.c               | 214 +++++++++++++++++++++++++
 security/selinux/include/objsec.h      |  14 +-
 security/selinux/include/security.h    |  10 ++
 security/selinux/selinuxfs.c           |  20 +++
 8 files changed, 274 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/configs/openeuler_defconfig b/arch/arm64/configs/openeuler_defconfig
index 4c23df4afee2..505f822d7b55 100644
--- a/arch/arm64/configs/openeuler_defconfig
+++ b/arch/arm64/configs/openeuler_defconfig
@@ -6499,6 +6499,7 @@ CONFIG_SECURITY_SELINUX_AVC_STATS=y
 CONFIG_SECURITY_SELINUX_CHECKREQPROT_VALUE=1
 CONFIG_SECURITY_SELINUX_SIDTAB_HASH_BITS=9
 CONFIG_SECURITY_SELINUX_SID2STR_CACHE_SIZE=256
+CONFIG_SECURITY_SELINUX_GENFSCON_NS=y
 # CONFIG_SECURITY_SMACK is not set
 # CONFIG_SECURITY_TOMOYO is not set
 CONFIG_SECURITY_APPARMOR=y
diff --git a/arch/x86/configs/openeuler_defconfig b/arch/x86/configs/openeuler_defconfig
index 40f6d273147c..57d0871a4cdf 100644
--- a/arch/x86/configs/openeuler_defconfig
+++ b/arch/x86/configs/openeuler_defconfig
@@ -7590,6 +7590,7 @@ CONFIG_SECURITY_SELINUX_AVC_STATS=y
 CONFIG_SECURITY_SELINUX_CHECKREQPROT_VALUE=1
 CONFIG_SECURITY_SELINUX_SIDTAB_HASH_BITS=9
 CONFIG_SECURITY_SELINUX_SID2STR_CACHE_SIZE=256
+CONFIG_SECURITY_SELINUX_GENFSCON_NS=y
 # CONFIG_SECURITY_SMACK is not set
 # CONFIG_SECURITY_TOMOYO is not set
 CONFIG_SECURITY_APPARMOR=y
diff --git a/security/security.c b/security/security.c
index 3c26d412acd2..0e545c5de07c 100644
--- a/security/security.c
+++ b/security/security.c
@@ -1006,6 +1006,7 @@ static void inode_free_by_rcu(struct rcu_head *head)
 void security_inode_free(struct inode *inode)
 {
 	integrity_inode_free(inode);
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	/*
 	 * The SELinux Genfscon Label Namespace feature involves another
 	 * call_rcu() in selinux_inode_free_security(), which must be called
@@ -1014,6 +1015,7 @@ void security_inode_free(struct inode *inode)
 	 * the same per-cpu list.
 	 */
 	preempt_disable();
+#endif
 	call_void_hook(inode_free_security, inode);
 	/*
 	 * The inode may still be referenced in a path walk and
@@ -1027,7 +1029,9 @@ void security_inode_free(struct inode *inode)
 	if (inode->i_security)
 		call_rcu((struct rcu_head *)inode->i_security,
 				inode_free_by_rcu);
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	preempt_enable();
+#endif
 }
 
 int security_dentry_init_security(struct dentry *dentry, int mode,
diff --git a/security/selinux/Kconfig b/security/selinux/Kconfig
index 9e921fc72538..9bdf17a1a278 100644
--- a/security/selinux/Kconfig
+++ b/security/selinux/Kconfig
@@ -115,3 +115,15 @@ config SECURITY_SELINUX_SID2STR_CACHE_SIZE
 	  conversion.  Setting this option to 0 disables the cache completely.
 
 	  If unsure, keep the default value.
+
+config SECURITY_SELINUX_GENFSCON_NS
+	bool "SELinux Genfscon Label Namespace"
+	depends on SECURITY_SELINUX
+	default n
+	help
+	  This option enables SELinux Genfscon Label Namespace, which is a
+	  feature that supports creating namespaces for file labels on some
+	  filesystems that only support the "genfscon" file labeling statement.
+	  Currently only procfs and sysfs are supported.
+
+	  If you don't need this feature, answer N.
diff --git a/security/selinux/hooks.c b/security/selinux/hooks.c
index 899ddf9ccd4f..220c4d26b4b7 100644
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c
@@ -109,9 +109,11 @@ struct selinux_state selinux_state;
 /* SECMARK reference count */
 static atomic_t selinux_secmark_refcount = ATOMIC_INIT(0);
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 const char *SELINUX_NS_INIT_NAME = "";
 static struct selinux_ns init_selinux_ns;
 static struct kmem_cache *inode_security_cache;
+#endif
 
 #ifdef CONFIG_SECURITY_SELINUX_DEVELOP
 static int selinux_enforcing_boot __initdata;
@@ -156,6 +158,7 @@ static int __init checkreqprot_setup(char *str)
 }
 __setup("checkreqprot=", checkreqprot_setup);
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 bool selinux_ns_enabled;
 
 enum sel_ns_fs_type {
@@ -249,6 +252,7 @@ static struct inode_security_struct *isec_alloc(struct inode *inode,
 
 	return isec;
 }
+#endif
 
 /**
  * selinux_secmark_enabled - Check to see if SECMARK is currently enabled
@@ -314,7 +318,9 @@ static void cred_init_security(void)
 
 	tsec = selinux_cred(cred);
 	tsec->osid = tsec->sid = SECINITSID_KERNEL;
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	tsec->ns = &init_selinux_ns;
+#endif
 }
 
 /*
@@ -343,6 +349,7 @@ static inline u32 task_sid(const struct task_struct *task)
 
 static int inode_doinit_with_dentry(struct inode *inode, struct dentry *opt_dentry);
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 static struct inode_security_struct *find_isec_ns(struct inode *inode,
 						  bool may_sleep)
 {
@@ -505,6 +512,104 @@ static void inode_free_security(struct inode *inode)
 	 */
 	call_rcu(&isech->rcu, isech_free_rcu);
 }
+#else
+/*
+ * Try reloading inode security labels that have been marked as invalid.  The
+ * @may_sleep parameter indicates when sleeping and thus reloading labels is
+ * allowed; when set to false, returns -ECHILD when the label is
+ * invalid.  The @dentry parameter should be set to a dentry of the inode.
+ */
+static int __inode_security_revalidate(struct inode *inode,
+				       struct dentry *dentry,
+				       bool may_sleep)
+
+{
+	struct inode_security_struct *isec = selinux_inode(inode);
+
+	might_sleep_if(may_sleep);
+
+	if (selinux_initialized(&selinux_state) &&
+	    isec->initialized != LABEL_INITIALIZED) {
+		if (!may_sleep)
+			return -ECHILD;
+
+		/*
+		 * Try reloading the inode security label.  This will fail if
+		 * @opt_dentry is NULL and no dentry for this inode can be
+		 * found; in that case, continue using the old label.
+		 */
+		inode_doinit_with_dentry(inode, dentry);
+	}
+	return 0;
+}
+
+static struct inode_security_struct *inode_security_novalidate(struct inode *inode)
+{
+	return selinux_inode(inode);
+}
+
+static struct inode_security_struct *inode_security_rcu(struct inode *inode, bool rcu)
+{
+	int error;
+
+	error = __inode_security_revalidate(inode, NULL, !rcu);
+	if (error)
+		return ERR_PTR(error);
+	return selinux_inode(inode);
+}
+
+/*
+ * Get the security label of an inode.
+ */
+static struct inode_security_struct *inode_security(struct inode *inode)
+{
+	__inode_security_revalidate(inode, NULL, true);
+	return selinux_inode(inode);
+}
+
+static struct inode_security_struct *backing_inode_security_novalidate(struct dentry *dentry)
+{
+	struct inode *inode = d_backing_inode(dentry);
+
+	return selinux_inode(inode);
+}
+
+/*
+ * Get the security label of a dentry's backing inode.
+ */
+static struct inode_security_struct *backing_inode_security(struct dentry *dentry)
+{
+	struct inode *inode = d_backing_inode(dentry);
+
+	__inode_security_revalidate(inode, dentry, true);
+	return selinux_inode(inode);
+}
+
+static void inode_free_security(struct inode *inode)
+{
+	struct inode_security_struct *isec = selinux_inode(inode);
+	struct superblock_security_struct *sbsec;
+
+	if (!isec)
+		return;
+	sbsec = inode->i_sb->s_security;
+	/*
+	 * As not all inode security structures are in a list, we check for
+	 * empty list outside of the lock to make sure that we won't waste
+	 * time taking a lock doing nothing.
+	 *
+	 * The list_del_init() function can be safely called more than once.
+	 * It should not be possible for this function to be called with
+	 * concurrent list_add(), but for better safety against future changes
+	 * in the code, we use list_empty_careful() here.
+	 */
+	if (!list_empty_careful(&isec->list)) {
+		spin_lock(&sbsec->isec_lock);
+		list_del_init(&isec->list);
+		spin_unlock(&sbsec->isec_lock);
+	}
+}
+#endif
 
 static void superblock_free_security(struct super_block *sb)
 {
@@ -755,12 +860,14 @@ static int parse_sid(struct super_block *sb, const char *s, u32 *sid)
 	return rc;
 }
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 static bool support_selinux_ns(struct super_block *sb)
 {
 	int idx = fs_str_idx(sb->s_type->name);
 
 	return idx != NS_TYPE_NUM && selinux_ns_fs_enabled[idx];
 }
+#endif
 
 /*
  * Allow filesystems with binary mount data to explicitly set mount point
@@ -816,8 +923,10 @@ static int selinux_set_mnt_opts(struct super_block *sb,
 	    && !opts)
 		goto out;
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	if (selinux_ns_enabled && support_selinux_ns(sb))
 		sbsec->flags |= SE_SBSUPPORT_NS;
+#endif
 
 	root_isec = backing_inode_security_novalidate(root);
 
@@ -1464,6 +1573,7 @@ static inline u16 socket_type_to_security_class(int family, int type, int protoc
 	return SECCLASS_SOCKET;
 }
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 static int selinux_genfs_get_sid(struct dentry *dentry,
 				 u16 tclass,
 				 u16 flags,
@@ -1518,6 +1628,45 @@ static int selinux_genfs_get_sid(struct dentry *dentry,
 out:
 	return rc;
 }
+#else
+static int selinux_genfs_get_sid(struct dentry *dentry,
+				 u16 tclass,
+				 u16 flags,
+				 u32 *sid)
+{
+	int rc;
+	struct super_block *sb = dentry->d_sb;
+	char *buffer, *path;
+
+	buffer = (char *)__get_free_page(GFP_KERNEL);
+	if (!buffer)
+		return -ENOMEM;
+
+	path = dentry_path_raw(dentry, buffer, PAGE_SIZE);
+	if (IS_ERR(path))
+		rc = PTR_ERR(path);
+	else {
+		if (flags & SE_SBPROC) {
+			/* each process gets a /proc/PID/ entry. Strip off the
+			 * PID part to get a valid selinux labeling.
+			 * e.g. /proc/1/net/rpc/nfs -> /net/rpc/nfs */
+			while (path[1] >= '0' && path[1] <= '9') {
+				path[1] = '/';
+				path++;
+			}
+		}
+		rc = security_genfs_sid(&selinux_state, sb->s_type->name,
+					path, tclass, sid);
+		if (rc == -ENOENT) {
+			/* No match in policy, mark as unlabeled. */
+			*sid = SECINITSID_UNLABELED;
+			rc = 0;
+		}
+	}
+	free_page((unsigned long)buffer);
+	return rc;
+}
+#endif
 
 static int inode_doinit_use_xattr(struct inode *inode, struct dentry *dentry,
 				  u32 def_sid, u32 *sid)
@@ -1584,7 +1733,11 @@ static int inode_doinit_use_xattr(struct inode *inode, struct dentry *dentry,
 static int inode_doinit_with_dentry(struct inode *inode, struct dentry *opt_dentry)
 {
 	struct superblock_security_struct *sbsec = NULL;
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	struct inode_security_struct *isec = inode_security_novalidate(inode);
+#else
+	struct inode_security_struct *isec = selinux_inode(inode);
+#endif
 	u32 task_sid, sid = 0;
 	u16 sclass;
 	struct dentry *dentry;
@@ -1710,8 +1863,13 @@ static int inode_doinit_with_dentry(struct inode *inode, struct dentry *opt_dent
 			 */
 			if (!dentry)
 				goto out_invalid;
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 			rc = selinux_genfs_get_sid(dentry, sclass,
 						   sbsec->flags, &sid, isec);
+#else
+			rc = selinux_genfs_get_sid(dentry, sclass,
+						   sbsec->flags, &sid);
+#endif
 			if (rc) {
 				dput(dentry);
 				goto out;
@@ -1842,7 +2000,11 @@ static int inode_has_perm(const struct cred *cred,
 		return 0;
 
 	sid = cred_sid(cred);
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	isec = inode_security_novalidate(inode);
+#else
+	isec = selinux_inode(inode);
+#endif
 
 	return avc_has_perm(&selinux_state,
 			    sid, isec->sid, isec->sclass, perms, adp);
@@ -3021,6 +3183,7 @@ static int selinux_fs_context_parse_param(struct fs_context *fc,
 
 /* inode security operations */
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 static int selinux_inode_alloc_security(struct inode *inode)
 {
 	struct inode_security_struct *isec;
@@ -3037,6 +3200,23 @@ static int selinux_inode_alloc_security(struct inode *inode)
 
 	return 0;
 }
+#else
+static int selinux_inode_alloc_security(struct inode *inode)
+{
+	struct inode_security_struct *isec = selinux_inode(inode);
+	u32 sid = current_sid();
+
+	spin_lock_init(&isec->lock);
+	INIT_LIST_HEAD(&isec->list);
+	isec->inode = inode;
+	isec->sid = SECINITSID_UNLABELED;
+	isec->sclass = SECCLASS_FILE;
+	isec->task_sid = sid;
+	isec->initialized = LABEL_INVALID;
+
+	return 0;
+}
+#endif
 
 static void selinux_inode_free_security(struct inode *inode)
 {
@@ -3105,9 +3285,13 @@ static int selinux_inode_init_security(struct inode *inode, struct inode *dir,
 
 	/* Possibly defer initialization to selinux_complete_init. */
 	if (sbsec->flags & SE_SBINITIALIZED) {
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 		struct inode_security_struct *isec;
 
 		isec = inode_security_novalidate(inode);
+#else
+		struct inode_security_struct *isec = selinux_inode(inode);
+#endif
 		isec->sclass = inode_mode_to_security_class(inode->i_mode);
 		isec->sid = newsid;
 		isec->initialized = LABEL_INITIALIZED;
@@ -3207,7 +3391,11 @@ static noinline int audit_inode_permission(struct inode *inode,
 					   int result)
 {
 	struct common_audit_data ad;
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	struct inode_security_struct *isec = inode_security_novalidate(inode);
+#else
+	struct inode_security_struct *isec = selinux_inode(inode);
+#endif
 	int rc;
 
 	ad.type = LSM_AUDIT_DATA_INODE;
@@ -4100,6 +4288,7 @@ static int selinux_task_alloc(struct task_struct *task,
 			    sid, sid, SECCLASS_PROCESS, PROCESS__FORK, NULL);
 }
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 static int selinux_cred_alloc_blank(struct cred *cred, gfp_t gfp)
 {
 	struct task_security_struct *tsec = selinux_cred(cred);
@@ -4117,6 +4306,7 @@ static void selinux_cred_free(struct cred *cred)
 
 	put_selinux_ns(tsec->ns);
 }
+#endif
 
 /*
  * prepare a new set of credentials for modification
@@ -4128,7 +4318,9 @@ static int selinux_cred_prepare(struct cred *new, const struct cred *old,
 	struct task_security_struct *tsec = selinux_cred(new);
 
 	*tsec = *old_tsec;
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	tsec->ns = get_selinux_ns(old_tsec->ns);
+#endif
 	return 0;
 }
 
@@ -4141,7 +4333,9 @@ static void selinux_cred_transfer(struct cred *new, const struct cred *old)
 	struct task_security_struct *tsec = selinux_cred(new);
 
 	*tsec = *old_tsec;
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	tsec->ns = get_selinux_ns(old_tsec->ns);
+#endif
 }
 
 static void selinux_cred_getsecid(const struct cred *c, u32 *secid)
@@ -4393,7 +4587,11 @@ static int selinux_task_kill(struct task_struct *p, struct kernel_siginfo *info,
 static void selinux_task_to_inode(struct task_struct *p,
 				  struct inode *inode)
 {
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	struct inode_security_struct *isec = inode_security_novalidate(inode);
+#else
+	struct inode_security_struct *isec = selinux_inode(inode);
+#endif
 	u32 sid = task_sid(p);
 
 	spin_lock(&isec->lock);
@@ -6730,7 +6928,11 @@ static void selinux_release_secctx(char *secdata, u32 seclen)
 
 static void selinux_inode_invalidate_secctx(struct inode *inode)
 {
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	struct inode_security_struct *isec = inode_security_novalidate(inode);
+#else
+	struct inode_security_struct *isec = selinux_inode(inode);
+#endif
 
 	spin_lock(&isec->lock);
 	isec->initialized = LABEL_INVALID;
@@ -7106,7 +7308,11 @@ static int selinux_lockdown(enum lockdown_reason what)
 struct lsm_blob_sizes selinux_blob_sizes __lsm_ro_after_init = {
 	.lbs_cred = sizeof(struct task_security_struct),
 	.lbs_file = sizeof(struct file_security_struct),
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	.lbs_inode = sizeof(struct inode_security_head),
+#else
+	.lbs_inode = sizeof(struct inode_security_struct),
+#endif
 	.lbs_ipc = sizeof(struct ipc_security_struct),
 	.lbs_msg_msg = sizeof(struct msg_security_struct),
 };
@@ -7270,8 +7476,10 @@ static struct security_hook_list selinux_hooks[] __lsm_ro_after_init = {
 	LSM_HOOK_INIT(file_open, selinux_file_open),
 
 	LSM_HOOK_INIT(task_alloc, selinux_task_alloc),
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	LSM_HOOK_INIT(cred_alloc_blank, selinux_cred_alloc_blank),
 	LSM_HOOK_INIT(cred_free, selinux_cred_free),
+#endif
 	LSM_HOOK_INIT(cred_prepare, selinux_cred_prepare),
 	LSM_HOOK_INIT(cred_transfer, selinux_cred_transfer),
 	LSM_HOOK_INIT(cred_getsecid, selinux_cred_getsecid),
@@ -7472,17 +7680,21 @@ static __init int selinux_init(void)
 	mutex_init(&selinux_state.status_lock);
 	mutex_init(&selinux_state.policy_mutex);
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	init_selinux_ns.name = SELINUX_NS_INIT_NAME;
 	refcount_set(&init_selinux_ns.count, 1);
+#endif
 
 	/* Set the security state for the initial task. */
 	cred_init_security();
 
 	default_noexec = !(VM_DATA_DEFAULT_FLAGS & VM_EXEC);
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	inode_security_cache = kmem_cache_create("selinux_inode_security",
 					    sizeof(struct inode_security_struct),
 					    0, SLAB_PANIC, NULL);
+#endif
 	avc_init();
 
 	avtab_cache_init();
@@ -7663,6 +7875,7 @@ int selinux_disable(struct selinux_state *state)
 }
 #endif
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 static void selinux_ns_free(struct work_struct *work)
 {
 	struct selinux_ns *ns = container_of(work, struct selinux_ns, work);
@@ -7703,3 +7916,4 @@ struct selinux_ns *selinux_ns_create(char *name)
 
 	return new_ns;
 }
+#endif
diff --git a/security/selinux/include/objsec.h b/security/selinux/include/objsec.h
index aa84b7d833b4..973b5aa18ded 100644
--- a/security/selinux/include/objsec.h
+++ b/security/selinux/include/objsec.h
@@ -35,7 +35,9 @@ struct task_security_struct {
 	u32 create_sid;		/* fscreate SID */
 	u32 keycreate_sid;	/* keycreate SID */
 	u32 sockcreate_sid;	/* fscreate SID */
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	struct selinux_ns *ns;	/* selinux namespace */
+#endif
 } __randomize_layout;
 
 enum label_initialized {
@@ -44,12 +46,14 @@ enum label_initialized {
 	LABEL_PENDING
 };
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 struct inode_security_head {
 	struct list_head head;	/* list head of inode_security_struct */
 	u32 task_sid;		/* the real SID of inode creater */
 	spinlock_t lock;
 	struct rcu_head rcu;	/* for freeing the inode_security_struct */
 };
+#endif
 
 struct inode_security_struct {
 	struct inode *inode;	/* back pointer to inode object */
@@ -59,8 +63,10 @@ struct inode_security_struct {
 	u16 sclass;		/* security class of this object */
 	unsigned char initialized;	/* initialization flag */
 	spinlock_t lock;
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	struct list_head isec_list;	/* list of isec */
 	struct selinux_ns *ns;	/* corresponding namespace */
+#endif
 };
 
 struct file_security_struct {
@@ -169,8 +175,12 @@ static inline struct file_security_struct *selinux_file(const struct file *file)
 	return file->f_security + selinux_blob_sizes.lbs_file;
 }
 
-static inline struct inode_security_head *selinux_inode(
-						const struct inode *inode)
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
+static inline struct inode_security_head *
+#else
+static inline struct inode_security_struct *
+#endif
+selinux_inode(const struct inode *inode)
 {
 	if (unlikely(!inode->i_security))
 		return NULL;
diff --git a/security/selinux/include/security.h b/security/selinux/include/security.h
index b9fd99d6bc23..995473ece47c 100644
--- a/security/selinux/include/security.h
+++ b/security/selinux/include/security.h
@@ -63,7 +63,9 @@
 #define SE_SBPROC		0x0200
 #define SE_SBGENFS		0x0400
 #define SE_SBGENFS_XATTR	0x0800
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 #define SE_SBSUPPORT_NS		0x1000
+#endif
 
 #define CONTEXT_STR	"context"
 #define FSCONTEXT_STR	"fscontext"
@@ -107,6 +109,7 @@ struct selinux_state {
 	struct mutex policy_mutex;
 } __randomize_layout;
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 extern const char *SELINUX_NS_INIT_NAME;
 
 struct selinux_ns {
@@ -117,13 +120,18 @@ struct selinux_ns {
 };
 
 #define current_selinux_ns ((selinux_cred(current_cred()))->ns)
+#endif
 
 void selinux_avc_init(struct selinux_avc **avc);
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 struct selinux_ns *selinux_ns_create(char *name);
 void put_selinux_ns(struct selinux_ns *ns);
+#endif
 
 extern struct selinux_state selinux_state;
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 extern bool selinux_ns_enabled;
+#endif
 
 static inline bool selinux_initialized(const struct selinux_state *state)
 {
@@ -474,7 +482,9 @@ extern void avtab_cache_init(void);
 extern void ebitmap_cache_init(void);
 extern void hashtab_cache_init(void);
 extern int security_sidtab_hash_stats(struct selinux_state *state, char *page);
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 extern struct inode_security_struct *inode_security_novalidate(
 		struct inode *inode);
+#endif
 
 #endif /* _SELINUX_SECURITY_H_ */
diff --git a/security/selinux/selinuxfs.c b/security/selinux/selinuxfs.c
index 2ba2da4461e7..2ed7524c1469 100644
--- a/security/selinux/selinuxfs.c
+++ b/security/selinux/selinuxfs.c
@@ -63,7 +63,9 @@ enum sel_inos {
 	SEL_STATUS,	/* export current status using mmap() */
 	SEL_POLICY,	/* allow userspace to read the in kernel policy */
 	SEL_VALIDATE_TRANS, /* compute validatetrans decision */
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	SEL_NS,		/* namespace */
+#endif
 	SEL_INO_NEXT,	/* The next inode number to use */
 };
 
@@ -334,6 +336,7 @@ static const struct file_operations sel_disable_ops = {
 	.llseek		= generic_file_llseek,
 };
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 static ssize_t sel_write_ns(struct file *file, const char __user *buf,
 			    size_t count, loff_t *ppos)
 {
@@ -404,6 +407,7 @@ static const struct file_operations sel_ns_ops = {
 	.read	= sel_read_ns,
 	.llseek	= generic_file_llseek,
 };
+#endif
 
 static ssize_t sel_read_policyvers(struct file *filp, char __user *buf,
 				   size_t count, loff_t *ppos)
@@ -1530,7 +1534,11 @@ static int sel_make_bools(struct selinux_policy *newpolicy, struct dentry *bool_
 			goto out;
 		}
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 		isec = inode_security_novalidate(inode);
+#else
+		isec = selinux_inode(inode);
+#endif
 		ret = selinux_policy_genfs_sid(newpolicy, "selinuxfs", page,
 					 SECCLASS_FILE, &sid);
 		if (ret) {
@@ -2129,7 +2137,11 @@ static int sel_fill_super(struct super_block *sb, struct fs_context *fc)
 	struct inode *inode;
 	struct inode_security_struct *isec;
 
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	static struct tree_descr selinux_files[] = {
+#else
+	static const struct tree_descr selinux_files[] = {
+#endif
 		[SEL_LOAD] = {"load", &sel_load_ops, S_IRUSR|S_IWUSR},
 		[SEL_ENFORCE] = {"enforce", &sel_enforce_ops, S_IRUGO|S_IWUSR},
 		[SEL_CONTEXT] = {"context", &transaction_ops, S_IRUGO|S_IWUGO},
@@ -2149,11 +2161,15 @@ static int sel_fill_super(struct super_block *sb, struct fs_context *fc)
 		[SEL_POLICY] = {"policy", &sel_policy_ops, S_IRUGO},
 		[SEL_VALIDATE_TRANS] = {"validatetrans", &sel_transition_ops,
 					S_IWUGO},
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 		[SEL_NS] = {"ns", &sel_ns_ops, 0644},
+#endif
 		/* last one */ {""}
 	};
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	if (!selinux_ns_enabled)
 		selinux_files[SEL_NS].name = "";
+#endif
 
 	ret = selinux_fs_info_create(sb);
 	if (ret)
@@ -2184,7 +2200,11 @@ static int sel_fill_super(struct super_block *sb, struct fs_context *fc)
 	}
 
 	inode->i_ino = ++fsi->last_ino;
+#ifdef CONFIG_SECURITY_SELINUX_GENFSCON_NS
 	isec = inode_security_novalidate(inode);
+#else
+	isec = selinux_inode(inode);
+#endif
 	isec->sid = SECINITSID_DEVNULL;
 	isec->sclass = SECCLASS_CHR_FILE;
 	isec->initialized = LABEL_INITIALIZED;
-- 
2.17.1

